ICML-2026/paper-2619-gilbert-pollak/best_code/pipeline/geosteiner-5.3/lp_solve_2.3/lp_solve.1
.TH LP_SOLVE 1
.SH NAME
lp_solve \- Solve (mixed integer) linear programming problem.
.SH SYNOPSIS
lp_solve [option]* "<" <input-file>
.SH OPTIONS
.TP 1.2i
-v
Verbose mode. Among other things, shows all the pivots.
.TP
-h
Help mode, prints the usage.
.TP
-d
Debug mode, all intermediate results are printed, and the branch-and-bound
decisions in case of (mixed) integer problems.
.TP
-min
minimize the objective function. This is the default for MPS input.
In lp_solve format you can specify minimization or maximization in the input
file as well. The command line option overrides.
.TP
-max
maximize the objective function. This is the default for lp_solve format
input.
In lp_solve format you can specify minimization or maximization in the input
file as well. The command line option overrides.
.TP
-p
Only functional for pure LP problems. Print the values of the dual
variables as well in the result. They are named r_1 until r_XXXXX unless
specified by the user.  Note that bounds (constraints on just one variable)
are not considered real constraints, and are not given a row in the matrix,
and are therefore not printed here.
.TP
-b <bound>
Specify an upper (when minimizing) or lower (when maximizing) limit for the
value of the objective function to
the program. Only useful for (mixed) integer problems.  If close enough, may
speed up the calculations. The same result can be obtained by adding an extra
constraint to the problem.
.TP
-c
When branching in MILP problems, take the ceiling of the selected non-integer
variable first instead of the floor. This can influence the speed of MILP
problems.
.TP
-e <value>
Specify the accuracy with which it is checked whether the value of a variable
is really integer. <value> must be between 0 and 0.5. Default value is 1e-6
and should be OK for most applications. Of course only useful for MILP
problems.
.TP
-i
Print all intermediate valid solutions. Can give you useful
solutions even if the total run time is too long.
Only useful for (mixed) integer problems.
.TP
-s
Both rows and columns are scaled according to the geometric mean of the
coefficients on them before solving. This might improve the numerical
stability of your problem.
.TP
-I
Print info after reinverting.
.TP
-t
Trace pivot selection.
.TP
-mps
Read from MPS file instead of lp file. For a short introduction to MPS see
ftp://softlib.cs.rice.edu/pub/miplib/mps_format.
.TP
-degen
Use random perturbations to reduce degeneracy, can increase numerical
instability.
.SH DESCRIPTION
The linear programming problem can be formulated as: Solve A.x >= V1, with
V2.x maximal. A is a matrix, x a vector of (nonnegative) variables, V1 a
vector called the right hand side, and V2 a vector specifying the objective
function.
.br
Any number of the variables may be specified to be of type integer.
.br
This program solves problems of this kind. It is slightly more general than
the above problem, in that every row of A (specifying one constraint) can have
its own (in)equality, <=, >= or =. The result specifies values for all
variables.
.br
Uses a 'Simplex' algorithm and sparse matrix methods, for pure LP problems.
If one or more of the variables is declared integer, the Simplex algorithm is
iterated with a branch and bound algorithm, until the desired optimal
solution is found.
.br
The "-i" option will print all intermediate valid solutions.
.SH "INPUT SYNTAX"
The default input syntax is a set of algebraic expressions and "int"
declarations in the following order:
.sp
<objective function>
.br
<constraint>+
.br
<declaration>*
.sp
where:
.TP 0.2i
-
<objective function> is a linear combination of variables, ending with a
semicolon, optionally preceded by "max: " or "min: " to indicate whether you
want it to be minimized or maximized. The case is not important, "Max:" or
"MAX:" will work as well. Maximization is the default.
.TP
-
<constraint> is an optional constraint name followed by a colon plus a
linear combination of variables and constants, followed by a relational
operator, followed again by a linear combination of variables and constants,
ending with a semicolon. The relational operator can be any of the following:
"<" "<=" "=" ">" ">=". There is no semantic difference between "<" and "<="
nor between ">" and ">=" (even for integer variables!).
.TP
-
<declaration> is of the form: "int" <var>+ ";" Commas are allowed between
variables.
.sp
So, the simplest linear problem consists of an objective function and 1
constraint.
.SH EXAMPLE
The simple problem:
.sp
x1 >= 1
.br
x2 >= 1
.br
x1 + x2 >= 2
.br
minimize x1 + x2 (= maximize -(x1 + x2)), with x1 integer
.sp
can be written as follows:
.sp
-x1 + -x2;
.br
(or min: x1 + x2;)
.br
x1 > 1;
.br
x2 > 1;
.br
x1 + x2 > 2;
.br
int x1;
.sp
The correct result for (x1, x2) is of course (1, 1).
.br
With the -mps option, lp_solve will accept MPS as input format.
.SH BUGS
Specifying a constraint name for a bound (constraints on just single
variables) does not have an effect: they are not stored inside the main matrix
and are not assigned a dual variable.
.TP
-
The problem consists entirely of constraints on just single variables
(so-called "bounds", like x < 1; ) and no constraint with more than 1
variable (like x + 3 y > 17; ). This leaves lp_solve with an empty problem
matrix, as bounds are not stored in the main matrix. No real-life examples
should be of this form, so I am not really chasing this problem.
.TP
-
Many people forget that lp_solve can only handle POSITIVE values for the
variables. While reading MPS files it will however handle free or negative
variables by replacing them with a variable pair <var>_neg and <var>_pos or
-<var> respectively. It is up to the user to interpret the result of this
transformation.
.TP
- Sometimes problems are numerically unstable, and the unavoidable rounding
errors inside lp_solve will cause aborts. It is very hard to give general
solutions to this problem, but try to keep all values in your problem in the
order of magnitude of 1 by proper scaling. This is almost always better than
using lp_solves built-in scaling (with -s). Almost parallel constraints are
also not very good for numerical stability. Use "lp_solve -v" and observe the
values of the pivots to see if there are any dangerously large or low numbers
there.
.br
Building lp_solve with long doubles (see the Makefile) can help to increase
numerical stability, but will also increase the run time considerably.
.br
You can consult the author as well if you encounter numerical problems, but
please remember that it is very easy to formulate an infeasible LP problem, so
be sure there is a solution.
.SH SEE ALSO
The implementation of the simplex kernel was mainly based on:
.br
W. Orchard-Hays: "Advanced Linear Programming Computing Techniques",
McGraw-Hill 1968
.br
The mixed integer branch and bound part was inspired by:
.br
section 6.4 of "An Introduction to Linear Programming and Game Theory" by
Paul R. Thie, second edition published by John Wiley and Sons in 1988.
.br
This book refers to:
.br
Dakin, R.J., "A Tree Search Algorithm for MILP Problems", Comput. J., 8 (1965)
pp. 250-255
.SH ACKNOWLEDGEMENTS
The work of Jeroen Dirks made the transition from the basic version 1.5 to
the full version 2.0 possible. He contributed the procedural interface, a
built-in MPS reader, and many fixes and enhancements to the code.
.SH CONTRIBUTED BY
M.R.C.M. Berkelaar
.br
Eindhoven University of Technology
.br
Design Automation Section
.br
P.O. Box 513
.br
NL-5600 MB Eindhoven, The Netherlands
.br
phone +31-40-2474792
.br
E-mail: michel@es.ele.tue.nl
.SH STATUS
Use at own risk. Bug reports are welcome, as well as success stories.


ICML-2026/paper-2619-gilbert-pollak/best_code/pipeline/geosteiner-5.3/lp_solve_2.3/lp_solve.c
#include <string.h>
#include <time.h>
#include "lpkit.h"
#include "lpglob.h"
#include "patchlevel.h"

void print_help(char *argv[])
{
  printf("Usage of %s version %s:\n", argv[0], PATCHLEVEL);
  printf("%s [options] [[<]input_file]\n", argv[0]);
  printf("list of options:\n");
  printf("-h\t\tprints this message\n");
  printf("-v\t\tverbose mode, gives flow through the program\n");
  printf("-d\t\tdebug mode, all intermediate results are printed,\n\t\tand the branch-and-bound decisions\n");
  printf("-p\t\tprint the values of the dual variables\n");
  printf("-b <bound>\tspecify a lower bound for the objective function\n\t\tto the program. If close enough, may speed up the\n\t\tcalculations.\n");
  printf("-i\t\tprint all intermediate valid solutions.\n\t\tCan give you useful solutions even if the total run time\n\t\tis too long\n");
  printf("-e <number>\tspecifies the epsilon which is used to determine whether a\n\t\tfloating point number is in fact an integer.\n\t\tShould be < 0.5\n");
  printf("-c\t\tduring branch-and-bound, take the ceiling branch first\n");
  printf("-s\t\tuse automatic problem scaling.\n");  
  printf("-I\t\tprint info after reinverting\n");
  printf("-t\t\ttrace pivot selection\n");
  printf("-mps\t\tread from MPS file instead of lp file\n");
  printf("-degen\t\tuse perturbations to reduce degeneracy,\n\t\tcan increase numerical instability\n");
  printf("-trej <Trej>\tset minimum pivot value\n");
  printf("-parse_only\tparse input file but do not calculate (ie check)\n");
  printf("-time\t\tPrint CPU time to parse input and to calculate result\n");
  printf("-min\t\tMinimize the lp problem (overrules setting in file)\n");
  printf("-max\t\tMaximize the lp problem (overrules setting in file)\n");
}

void print_cpu_times(const char *info)
{
  static clock_t last_time = 0;
  clock_t new_time;

  new_time = clock();
  fprintf(stderr, "CPU Time for %s: %gs (%gs total since program start)\n",
	  info, (new_time - last_time) / (double) CLOCKS_PER_SEC,
	  new_time / (double) CLOCKS_PER_SEC);
  last_time = new_time;
}


int main(int argc, char *argv[])
{
  lprec *lp;
  int i;
  short verbose = FALSE;
  short debug = FALSE;
  short print_sol = FALSE;
  short print_duals = FALSE;
  short floor_first = TRUE;
  short scaling = FALSE;
  short print_at_invert = FALSE;
  short tracing = FALSE;
  short mps = FALSE;
  short anti_degen = FALSE;
  short print_timing = FALSE;
  short parse_only = FALSE;
  short do_presolve = FALSE;
  short objective = 0;
  int result;
  REAL obj_bound = (REAL)DEF_INFINITE;
  REAL epsilon = (REAL)DEF_EPSILON;
  FILE *fpin = stdin;

  Trej = 1e-4; /* default, MB */

  /* read command line arguments */

  for(i = 1; i < argc; i++) {
    if(strcmp(argv[i], "-v") == 0)
      verbose = TRUE;
    else if(strcmp(argv[i], "-d") == 0)
      debug = TRUE;
    else if(strcmp(argv[i], "-i") == 0)
      print_sol = TRUE;
    else if(strcmp(argv[i], "-c") == 0)
      floor_first = FALSE;
    else if(strcmp(argv[i], "-b") == 0)
      obj_bound = atof(argv[++i]);
    else if(strcmp(argv[i], "-e") == 0) {
      epsilon = atof(argv[++i]);
      if((epsilon <= 0.0) || (epsilon >= 0.5)) {
	fprintf(stderr, "Invalid epsilon %g; 0 < epsilon < 0.5\n",
		(double)epsilon);
	exit(EXIT_FAILURE);
      }
    }
    else if(strcmp(argv[i], "-p") == 0)
      print_duals = TRUE;
    else if(strcmp(argv[i], "-h") == 0) {
      print_help(argv); 
      exit(EXIT_SUCCESS);
    }
    else if(strcmp(argv[i], "-s") == 0)
      scaling = TRUE;
    else if(strcmp(argv[i], "-I") == 0)
      print_at_invert = TRUE;
    else if(strcmp(argv[i], "-t") == 0)
      tracing = TRUE;
    else if(strcmp(argv[i],"-mps") == 0)
      mps = TRUE;
    else if(strcmp(argv[i],"-degen") == 0)
      anti_degen = TRUE;
    else if(strcmp(argv[i],"-time") == 0) {
      if(clock() == -1)
	fprintf(stderr, "CPU times not available on this machine\n");
      else
	print_timing = TRUE;
    }
    else if(strcmp(argv[i],"-trej") == 0)
      Trej = atof(argv[++i]);
    else if(strcmp(argv[i],"-parse_only") == 0)
      /* only useful for parser software development */
      parse_only = TRUE;
    else if(strcmp(argv[i],"-presolve") == 0)
      do_presolve = TRUE;
    else if(strcmp(argv[i],"-min") == 0)
      objective = -1;
    else if(strcmp(argv[i],"-max") == 0)
      objective =  1;
    else if(fpin == stdin) {
      if((fpin = fopen(argv[i],"r")) == NULL) {
	fprintf(stderr,"Error, Unable to open input file '%s'\n",
		argv[i]);
	print_help(argv);
	exit(EXIT_FAILURE);
      }
    }
    else {
      fprintf(stderr, "Error, Unrecognized command line argument '%s'\n",
	      argv[i]);
      print_help(argv);
      exit(EXIT_FAILURE);
    }
  }

  if(mps)
    lp = read_mps(fpin, verbose);
  else /* standard lp_solve syntax expected */
    lp = read_lp_file(stdin, verbose, "lp" );

  if(fpin != stdin)
    fclose(fpin);

  if(print_timing)
    print_cpu_times("parsing input");

  if(parse_only)
    exit(0);

  if(verbose)
    print_lp(lp);

  if(scaling) {
    auto_scale(lp);
    if(verbose) {
      printf("Scaled matrix:\n");
      print_lp(lp);
    }
  }

  lp->print_sol = print_sol;
  lp->epsilon = epsilon;
  lp->print_duals = print_duals;
  lp->debug = debug;
  lp->floor_first = floor_first;
  lp->print_at_invert = print_at_invert;
  lp->trace = tracing;
  if(obj_bound != DEF_INFINITE)
    lp->obj_bound = obj_bound;
  lp->anti_degen = anti_degen;
  lp->do_presolve = do_presolve;

  if(verbose) {
    printf("Solving\n");
    lp->verbose = TRUE;
  }

  result = solve(lp);

  if(print_timing)
    print_cpu_times("solving");

  if(result == OPTIMAL) {
    print_solution(lp);
    if(verbose)
      fprintf(stderr,
	      "Branch & Bound depth: %d\nNodes processed: %d\nSimplex pivots: %d\n",
	      lp->max_level, lp->total_nodes, lp->total_iter);
  }

  else if(result == INFEASIBLE)
    printf("This problem is infeasible\n");
  else if(result == UNBOUNDED)
    printf("This problem is unbounded\n");
  else if(result == FAILURE)
    printf("lp_solve failed\n");
  return(result);
}
